### ISSUE TITLE
Some questions about your project

### ISSUE THREAD

**Issue by imetallica** [jcclinton/wow-emulator#1]
Hello, I know your project is stopped, but can you give me a few hints:
- Have you done some metrics on speed/memory comparison with CMaNGOS or Trinity? I know this is very much useless but, damn, I'm curious! :-)
- I had the damn same idea as you, but I was thinking on writing in Elixir instead. Do you believe it is doable to be done/spend time on it?

I used to develop for a network MaNGOS a long time ago. Hell, only if I knew Erlang at that time, maybe it could have been another viable solution.

**Comment by jcclinton** [comment]
I never ran any performance testing, but my emulator is almost definitely
slower than mangos and trinity. Erlang is slower than C++ in general, and I
never did any performance optimization. If mine was faster, it would
probably be due to a lack of features rather than anything innate in
Erlang. I suspect it would still scale pretty well though.

I never had a problem with Erlangs syntax, so I never bothered to learn
Elixir.

**Comment by imetallica** [comment]
Yes I'm aware that it will be slower. But I believe that, depending on the amount of players and cores the machine have, it could reach about the same speed. Not to count the ability to do upgrades while running the server and the possibility to crash stuff and the server keep going. That would be cool to have. That's the why of my interest using the BEAM to make an alternative, even if it's just for the vanilla (which is the only wow I care, tbh).

I don't mind Erlang's syntax as well, but for me Ecto is the thing when we talk about dealing with databases, that's just the reason for me to use Elixir instead.

-----Mensagem Original-----
De: "Jamie Clinton" notifications@github.com
Enviada em: ‎25/‎12/‎2015 03:21
Para: "jcclinton/wow-emulator" wow-emulator@noreply.github.com
Cc: "Iuri L. Machado" iuri.machado@linuxmail.org
Assunto: Re: [wow-emulator] Some questions about your project (#1)

I never ran any performance testing, but my emulator is almost definitely
slower than mangos and trinity. Erlang is slower than C++ in general, and I
never did any performance optimization. If mine was faster, it would
probably be due to a lack of features rather than anything innate in
Erlang. I suspect it would still scale pretty well though.

I never had a problem with Erlangs syntax, so I never bothered to learn
Elixir.

**Comment by jcclinton** [comment]
I agree, i think it would scale very well also. I suspect it is easy to
catch bottlenecks also, but ive never delved too deeply into the debug
tools.

You can run Elixir modules alongside Erlang modules, iirc. Personally I
thought Erlang was a great syntax for the gen behavior modules.

Are you planning on actually building a vanilla wow emulator from scratch?
The biggest difficulty is the lack of any documentation on the
client/server api. Also, Erlang uses the opposite endianness as C++. The
state updates to the clients are a beast, but once you learn it, its easy.

**Comment by imetallica** [comment]
Seeing your project made me even more excited to build it. I've already started, by the realm server implementation.

Yes I know about the interoperability between the two, but well, since I'm going to use Ecto I've decided to write everything in Elixir. And Protocols summed with metaprogramming caught my eye.

Also, there is an problem with high load and LoS/collision that I think Erlang can solve, by distributing the computations. Easy to implement with what we have.

-----Mensagem Original-----
De: "Jamie Clinton" notifications@github.com
Enviada em: ‎25/‎12/‎2015 14:24
Para: "jcclinton/wow-emulator" wow-emulator@noreply.github.com
Cc: "Iuri L. Machado" iuri.machado@linuxmail.org
Assunto: Re: [wow-emulator] Some questions about your project (#1)

I agree, i think it would scale very well also. I suspect it is easy to
catch bottlenecks also, but ive never delved too deeply into the debug
tools.

You can run Elixir modules alongside Erlang modules, iirc. Personally I
thought Erlang was a great syntax for the gen behavior modules.

Are you planning on actually building a vanilla wow emulator from scratch?
The biggest difficulty is the lack of any documentation on the
client/server api. Also, Erlang uses the opposite endianness as C++. The
state updates to the clients are a beast, but once you learn it, its easy.

**Comment by jcclinton** [comment]
I think the high load could be solved by the distributed nature of Erlang;
that would be a fun and interesting problem to solve. For los/collision I
was thinking, if Erlang were too slow, to use a C node. That way you could
pass off processing intense tasks to C/C++.

**Comment by imetallica** [comment]
Indeed it is a very interesting problem to solve. And I think it's a shame the beam ecosystem is not widely used. It would be a very perfect match. And not only that, the tooling, like building a web service to monitor the server in real time, a very cool almost "real time" Armoury. O well, I'm thinking too much into the possible future.

Also, I've been thinking distributing "worlds" to servers, like, one server for Kalimdor, one for Eastern Kingdoms. Hell so many cool alternatives. :-)

**Comment by imetallica** [comment]
The problem with collision/LoS is the mmaps calculations. I was discussing this with a friend that took time to do a full sweep into MaNGOS and he discovered that these calculations take too long to be completed. What possibly blizzard does is, distribute the load to another server. This is why it lags alot when servers have like, 4k+ people. The internal latency is due to the time it takes for completion of these calculations.

**Comment by jcclinton** [comment]
Thats why I thought a C port would be a good method to handle the mmaps
calculatons.

Anyway, if you have any questions feel free to ask. If you ever get a
server working, let me know! good luck

**Comment by imetallica** [comment]
@jcclinton Thanks, I'll let you know. :-)

**Comment by imetallica** [comment]
Hey @jcclinton, I've begun!

https://github.com/imetallica/pyrite
